src/sentry/api/endpoints/organization_traces.py

```diff
@@ -0,0 +1,160 @@
+from collections import defaultdict
+from collections.abc import Mapping
+from datetime import datetime, timedelta
+from typing import Any, TypedDict, cast
+
+from rest_framework import serializers
+from rest_framework.exceptions import ParseError
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from sentry import options
+from sentry.api.api_owners import ApiOwner
+from sentry.api.api_publish_status import ApiPublishStatus
+from sentry.api.base import region_silo_endpoint
+from sentry.api.bases import NoProjects, OrganizationEventsV2EndpointBase
+from sentry.api.paginator import GenericOffsetPaginator
+from sentry.api.utils import handle_query_errors
+from sentry.models.organization import Organization
+from sentry.search.events.builder import SpansIndexedQueryBuilder
+from sentry.search.events.types import ParamsType, QueryBuilderConfig
+from sentry.snuba.dataset import Dataset
+from sentry.snuba.referrer import Referrer
+
+
+class TraceResult(TypedDict):
+    trace: str
+    numSpans: int
+    name: str | None
+    duration: int
+    spans: list[Mapping[str, Any]]
+
+
+class OrganizationTracesSerializer(serializers.Serializer):
+    field = serializers.ListField(required=True, allow_empty=False, child=serializers.CharField())
+    query = serializers.CharField(required=False)
+    maxSpansPerTrace = serializers.IntegerField(default=1, min_value=1, max_value=100)
+
+
+@region_silo_endpoint
+class OrganizationTracesEndpoint(OrganizationEventsV2EndpointBase):
+    publish_status = {
+        "GET": ApiPublishStatus.EXPERIMENTAL,
+    }
+    owner = ApiOwner.PERFORMANCE
+
+    def get(self, request: Request, organization: Organization) -> Response:
+        try:
+            snuba_params, params = self.get_snuba_dataclass(request, organization)
+        except NoProjects:
+            return Response(status=404)
+
+        serializer = OrganizationTracesSerializer(data=request.GET)
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=400)
+        serialized = serializer.validated_data
+
+        per_page = self.get_per_page(request)
+        limit = per_page * serialized["maxSpansPerTrace"]
+
+        # Abort early if number of trace x spans per trace
+        # exceeds 10_000 (the snuba query limit),
+        if limit > 10_000:
+            raise ParseError(detail="Too many spans per trace.")
+
```

Comment:
I suppose this is possible but you could just limit `maxSpansPerTrace` to 90 or lower since we don't need a ton of them anyway (we can just say "see more in trace"). Mixed with per_page of 100 that means it's impossible for the query limit to be hit.

Reply:
ahh i forgot to remove this, we have a limit of 100 for per page and max spans per trace, so it'll always be under 10k if it passes validation